File: code_files_MATLAB/LikelihoodFunc.m
function [sumloglik,logLik,a,nua,LTMpifinal,stdeps]=LikelihoodFunc(inflation,fedfunds,param)

global Delta

n=length(inflation);

sigmapi=abs(param(1,1));
pibar=param(2,1); 
lambdapi=abs(param(3,1));
sigmaa=abs(param(4,1));
lambdaa=abs(param(5,1));
abar=0;
rNbar=mean(fedfunds);

%initial values
a(1,1)=abar;
nua(1,1)=0.5*sigmaa^2/(2*lambdaa);

for i=1:n-1,   
    %Approximate solution of Ornstein-Uhlenbeck (OU)
    LTMpi(i,1)=pibar-a(i,1)*(fedfunds(i,1)-rNbar);
    eps(i,1)=inflation(i+1,1)-exp(-lambdapi*Delta)*inflation(i,1)-LTMpi(i,1)*(1-exp(-lambdapi*Delta));
    vareps=sigmapi^2/(2*lambdapi)*(1-exp(-2*lambdapi*Delta));
    stdeps(i,1)=eps(i,1)/sqrt(vareps);

    %update a and nua
    vara = ((rNbar-fedfunds(i,1))*lambdapi*nua(i,1)/sigmapi)^2/(2*lambdaa)*(1-exp(-2*lambdaa*Delta)); 
    a(i+1,1)=exp(-lambdaa*Delta)*a(i,1)+abar*(1-exp(-lambdaa*Delta))+sign(rNbar-fedfunds(i,1))*sqrt(vara)*stdeps(i,1);    
       
    nua(i+1,1)=max(0,nua(i,1)+(sigmaa^2-2*lambdaa*nua(i,1)-(rNbar-fedfunds(i,1))^2*lambdapi^2*nua(i,1)^2/sigmapi^2)*Delta);
    
    % normal likelihood for whole time period
    epsVec=[eps(i,1)];
    VarMat=[vareps];
    Nv=length(epsVec);

    logLik(i,1)=log(1./((2.*pi)^(Nv/2).*sqrt(det(VarMat))))-1/2*(epsVec*(VarMat\epsVec'));
end

if mean(LTMpi)+2*std(LTMpi)>quantile(inflation,0.9) || mean(LTMpi)-2*std(LTMpi)<quantile(inflation,0.1) || mean(LTMpi)>mean(inflation)+0.0005 || mean(LTMpi)<mean(inflation)-0.0005 || lambdaa<0.5
    pen=10^10;
else
    pen=0;
end

sumloglik=-sum(logLik(2:end,1))+pen; % this the sum of log likelihood (in negative sign since fminsearch minimizes)

LTMpifinal=pibar-a.*(fedfunds-rNbar);

end

File: code_files_R/LikelihoodFunc.R
LikelihoodFunc <- function(inflation, fedfunds, param) {
  # --------------------------------------------------------------
  # LikelihoodFunc:
  #   Computes the (negative) log-likelihood for the inflation block
  #   based on an Ornstein-Uhlenbeck process and a Kalman learning model.
  #
  # Inputs:
  #   inflation - observed inflation time series
  #   fedfunds  - Federal funds rate series
  #   param     - parameter vector (sigmapi, pibar, lambdapi, sigmaa, lambdaa)
  #
  # Output (list):
  #   sumloglik - negative total log-likelihood (scalar)
  #   logLik    - vector of log-likelihood values per observation
  #   a         - learning coefficient sequence
  #   nua       - uncertainty sequence
  #   LTMpifinal - long-term mean of inflation
  #   stdeps    - standardized residuals
  # --------------------------------------------------------------
  
  # Use global Delta (time step)
  Delta <- get("Delta", envir = .GlobalEnv)
  
  n <- length(inflation)
  
  # Extract parameters (ensure nonnegativity where required)
  sigmapi <- abs(param[1])
  pibar   <- param[2]
  lambdapi <- abs(param[3])
  sigmaa  <- abs(param[4])
  lambdaa <- abs(param[5])
  abar <- 0
  rNbar <- mean(fedfunds)
  
  # Initialize state variables
  a <- numeric(n)
  nua <- numeric(n)
  a[1] <- abar
  nua[1] <- 0.5 * sigmaa^2 / (2 * lambdaa)
  
  # Initialize storage vectors
  LTMpi <- numeric(n - 1)
  eps <- numeric(n - 1)
  stdeps <- numeric(n - 1)
  logLik <- numeric(n - 1)
  
  # Main likelihood loop
  for (i in 1:(n - 1)) {
    # Approximate solution of Ornstein-Uhlenbeck (OU)
    LTMpi[i] <- pibar - a[i] * (fedfunds[i] - rNbar)
    eps[i] <- inflation[i + 1] -
      exp(-lambdapi * Delta) * inflation[i] -
      LTMpi[i] * (1 - exp(-lambdapi * Delta))
    vareps <- sigmapi^2 / (2 * lambdapi) * (1 - exp(-2 * lambdapi * Delta))
    stdeps[i] <- eps[i] / sqrt(vareps)
    
    # Update a and nua
    vara <- ((rNbar - fedfunds[i]) * lambdapi * nua[i] / sigmapi)^2 /
      (2 * lambdaa) * (1 - exp(-2 * lambdaa * Delta))
    a[i + 1] <- exp(-lambdaa * Delta) * a[i] +
      abar * (1 - exp(-lambdaa * Delta)) +
      sign(rNbar - fedfunds[i]) * sqrt(vara) * stdeps[i]
    
    nua[i + 1] <- max(
      0,
      nua[i] + (sigmaa^2 -
                  2 * lambdaa * nua[i] -
                  (rNbar - fedfunds[i])^2 * lambdapi^2 * nua[i]^2 / sigmapi^2) * Delta
    )
    
    # Normal log-likelihood for this observation
    logLik[i] <- log(1 / (sqrt(2 * pi * vareps))) - 0.5 * (eps[i]^2 / vareps)
  }
  
  # Penalty conditions (same logic as MATLAB)
  if (
    mean(LTMpi) + 2 * sd(LTMpi) > quantile(inflation, 0.9) ||
    mean(LTMpi) - 2 * sd(LTMpi) < quantile(inflation, 0.1) ||
    mean(LTMpi) > mean(inflation) + 0.0005 ||
    mean(LTMpi) < mean(inflation) - 0.0005 ||
    lambdaa < 0.5
  ) {
    pen <- 1e10
  } else {
    pen <- 0
  }
  
  # Sum of log-likelihood (negative for optimization)
  sumloglik <- -sum(logLik[-1]) + pen
  
  # Compute final long-term inflation mean
  LTMpifinal <- pibar - a * (fedfunds - rNbar)
  
  # Return list
  return(list(
    sumloglik = sumloglik,
    logLik = logLik,
    a = a,
    nua = nua,
    LTMpifinal = LTMpifinal,
    stdeps = stdeps
  ))
}
